nucypher/blockchain/eth/agents.py

```diff
@@ -396,6 +392,51 @@ def deposit_tokens(self, staker_address: str, amount: int, lock_periods: int, se
         receipt = self.blockchain.send_transaction(contract_function=contract_function, sender_address=sender_address)
         return receipt
 
+    def construct_batch_deposit_parameters(self, deposits: Dict[str, List[Tuple[int, int]]]) -> Tuple[list, list, list, list]:
+        max_substakes = self.contract.functions.MAX_SUB_STAKES().call()
+        stakers, number_of_substakes, amounts, lock_periods = list(), list(), list(), list()
+        for staker, substakes in deposits.items():
+            if not 0 < len(substakes) <= max_substakes:
+                raise self.RequirementError(f"Number of substakes for staker {staker} must be >0 and ≤{max_substakes}")
+            # require(value >= minAllowableLockedTokens & & periods >= minLockedPeriods);
+            # require(info.value <= maxAllowableLockedTokens);
+            # require(info.subStakes.length == 0);
+            stakers.append(staker)
+            number_of_substakes.append(len(substakes))
+            staker_amounts, staker_periods = zip(*substakes)
+            amounts.extend(staker_amounts)
+            lock_periods.extend(staker_periods)
+
+        return stakers, number_of_substakes, amounts, lock_periods
+
+    @validate_checksum_address
+    def batch_deposit(self,
+                      stakers: list,
+                      number_of_substakes: list,
+                      amounts: list,
+                      lock_periods: list,
+                      sender_address: str,
+                      dry_run: bool = False,
+                      gas_limit: int = None) -> Union[dict, int]:
+        min_gas_batch_deposit = 250_000
+        if gas_limit and gas_limit < min_gas_batch_deposit:
+            raise ValueError(f"{gas_limit} is not enough gas for any batch deposit")
+
+        contract_function = self.contract.functions.batchDeposit(stakers, number_of_substakes, amounts, lock_periods)
+        if dry_run:
+            payload = {'from': sender_address}
+            if gas_limit:
+                payload['gas'] = gas_limit
+            estimated_gas = contract_function.estimateGas(payload)  # If TX is not correct, or there's not enough gas, this will fail.
+            if gas_limit and estimated_gas > gas_limit:
+                raise ValueError
```

Comment:
I guess some explanation in error will be helpful

Reply:
```suggestion
                raise ValueError(f"Estimated gas for transaction exceeds gas limit {gas_limit}")
```